### ISSUE TITLE
.TOML Contents Reset After Quiting from Menu or Shortcut - MacOS Sierra 10.12 

### ISSUE THREAD

**Issue by silvether** [demonsaw/demonsaw#27]
### **SEE COMMENTS FOR A MORE CURRENT UNDERSTANDING OF THE ISSUE**

**OS:** MacOS Sierra 10.12 (16A323)
**Demonsaw Version:** 64-bit OSX 3.1

**Expected Behaviour:**
When the save icon is initiated, Demonsaw saves current client settings to the .TOML and settings persist across restarts of the application.

**Current Behaviour:**
When the save icon is initiated, Demonsaw saves the current client settings, but when the app is exited, **the TOML file is overwritten to a default blank state.**

**Reproduction Steps:**

**Step 1:** Launch Demonsaw from within the original folder that is found in the [version 3.1 .tar.gz archive](http://demonsaw.com/download/3.1.0/demonsaw_osx_64.tar.gz)

![screen shot 2016-10-06 at 1 24 06 am](https://cloud.githubusercontent.com/assets/19334356/19143932/0c20cab8-8b64-11e6-9577-588b4f4dc197.png)

**Step 2:** Configure a client

<img width="592" alt="screen shot 2016-10-06 at 1 24 17 am" src="https://cloud.githubusercontent.com/assets/19334356/19143937/14e11bd0-8b64-11e6-897c-bfaf09afbffb.png">

**Step 3:** Press the save icon at the top left of Demonsaw to save the .TOML

<img width="113" alt="screen shot 2016-10-06 at 1 31 05 am" src="https://cloud.githubusercontent.com/assets/19334356/19144026/91dca230-8b64-11e6-884f-097912ee6283.png">

**Step 4:** Open the .TOML in a text editor such as Sublime and note that client configuration is present

<img width="868" alt="screen shot 2016-10-06 at 1 25 36 am" src="https://cloud.githubusercontent.com/assets/19334356/19143972/44ffcdac-8b64-11e6-8aa7-edcff8a7635e.png">

**Step 5:** Exit Demonsaw

<img width="211" alt="screen shot 2016-10-06 at 1 26 05 am" src="https://cloud.githubusercontent.com/assets/19334356/19143965/3c661bec-8b64-11e6-82ff-acc540261144.png">

**Step 6:** Note the now empty/default .TOML file, (the timestamp will confirm a change is committed upon quit)

<img width="868" alt="screen shot 2016-10-06 at 1 26 15 am" src="https://cloud.githubusercontent.com/assets/19334356/19143981/4f575b58-8b64-11e6-964f-d03d97bb28ba.png">

**Step 7:** Launch Demonsaw again from the same folder location to confirm that the vanilla .TOML file triggers the configuration dialogue. 

<img width="592" alt="screen shot 2016-10-06 at 1 24 17 am" src="https://cloud.githubusercontent.com/assets/19334356/19143986/549784e4-8b64-11e6-820b-791c25c961af.png">

**Additional Observations & Feedback:**

**Most Important:** The ability to save a client configuration is a core feature of Demonsaw and should be thoroughly tested before rtw. The current inability to persist settings across stop/starts **makes the OSX client unusable**, especially for groups that require unique configuration values.

**Less Important:** The vast majority of OSX applications do not require the .app to be rooted in a folder containing settings or configurations. It is common practice to allow the .app to reside in the Applications folder independently, while settings are stored in ~/Library, commonly ~/Library/Application Support/[NAME OF APP or ORG HERE]/ config files. This could even be accomplished by **creating a visible Demonsaw folder in the 'Documents' folder** where CSS and the .TOML could reside.(I find this more appropriate than ~/Library which is hidden by default)

_**I can understand how keeping configs in a folder and not scattering them around the OS, folders(hidden or visible) may be easier from a development perspective, but the current behaviour is confusing and obtuse to the average MacOS/OSX user._

Please let me know if I can provide any further clarification. I am also more than happy to assist in additional testing.

w/ love,
- silvether

__
@eijah @dev247 @AmbassadorP @camgaertner @Havatra

**Comment by camgaertner** [comment]
Great report. It appears that this sort of issue can occur when Demonsaw is not run under sudo privileges on OS X. 

Could you please attempt to reproduce this after starting Demonsaw via Terminal with the command 
`sudo ./demonsaw` ?

**Comment by silvether** [comment]
Thanks for the quick response Cam,

To test again with sudo I downloaded a fresh copy of 'demonsaw_osx_64.tar.gz' from the website just to be sure. I also deleted associated files in ~/Library. 

**Files Removed:**

<img width="603" alt="screen shot 2016-10-06 at 1 10 53 pm" src="https://cloud.githubusercontent.com/assets/19334356/19167155/ab258104-8bc7-11e6-93bf-313dade11bb3.png">

**Vanilla Demonsaw Folder:**

![screen shot 2016-10-06 at 12 54 57 pm](https://cloud.githubusercontent.com/assets/19334356/19167188/c6d3a1ce-8bc7-11e6-99e0-78b111cbbeab.png)

Now I cannot get Demonsaw to write a .TOML under either root or my non-privileged user. But even though the .TOML was written and then cleared on exit in my initial report, the results are the same.

_To add further validity to my test results, I just re-installed MacOS Sierra last night after wiping my disk completely._

Can you reproduce my results with a fresh download of Demonsaw?

**Comment by eijah** [comment]
Saving is different in OSX cause of the .app file.  Make sure that you have write access to the folder that contains demonsaw.app

-E

**Comment by silvether** [comment]
**Hypothesis:**

After chatting with Cam, we may have an idea as to why we run into issues with OSX and the need for root, in this case writing a .TOML file that is available to a standard OSX user. From my personal experience, OSX does not play well with running entire apps under the root user, especially GUIs. My understanding is the reason that we are required to use root for Demonsaw in certain parts is that *nix systems consider ports >1024 to be privileged. OSX given its BSD core maintains similar controls on low level ports. If Demonsaw does indeed require root privileges under OSX I think using `launchd` which runs under root by default may be a way to mitigate this issue, again hypothetical given my lack of access to code. There may even be another way aside from launchd outlined in the documentation I listed below.

**Here is what I was able to confirm in Apple's documentation:**

`Because launchd itself runs as the root user, if your only reason for using a privileged process is to run a daemon on a low-numbered port, you can let launchd open that port on your daemon’s behalf and pass the open socket to your daemon, thus eliminating the need for your code to run as the root user.` - [Elevating Privileges Safely](https://developer.apple.com/library/content/documentation/Security/Conceptual/SecureCodingGuide/Articles/AccessControl.html#//apple_ref/doc/uid/TP40002589-SW4#launchd)

I know this deal about port binding may seem off-topic but I think it might help with the potential issues we are seeing with root.

Thoughts?

**Comment by silvether** [comment]
Thanks @eijah, I can confirm by default that I have read/write access to the demonsaw folder recursively w/ the exception of eula.txt and version.txt(shouldn't matter).

<img width="483" alt="screen shot 2016-10-06 at 2 02 29 pm" src="https://cloud.githubusercontent.com/assets/19334356/19168546/cbcad408-8bcd-11e6-803a-f1f5e84d0ede.png">

<img width="493" alt="screen shot 2016-10-06 at 1 59 01 pm" src="https://cloud.githubusercontent.com/assets/19334356/19168401/3a8bf3be-8bcd-11e6-9e5a-6acee5aa32b3.png">

<img width="489" alt="screen shot 2016-10-06 at 1 58 26 pm" src="https://cloud.githubusercontent.com/assets/19334356/19168384/2a2816f6-8bcd-11e6-9200-f15e60237186.png">

I may set 777 permissions on the folder and its contents just for testing sake to see if the behaviour changes. 

**UPDATE:**

Performed `sudo chmod -R 777 demonsaw_osx_64/` successfully.

**Top Dir:**

`drwxrwxrwx@  6 silver  staff    204 Oct  6 14:16 demonsaw_osx_64`

**Contents:**

```
drwxrwxrwx@  3 silver  staff      102 Oct  2 18:29 demonsaw.app
-rwxrwxrwx@  1 silver  staff  2907296 Oct  2 18:47 demonsaw_cli
drwxrwxrwx@  8 silver  staff      272 Oct  2 17:44 theme
```

No save behaviour persists.

**Comment by silvether** [comment]
`drwxrwxrwx@  8 silver  staff    272 Oct  6 13:13 demonsaw_osx_64`

> drwx---rwx@  3 silver  staff      102 Oct  2 18:29 demonsaw.app
> -rwxrwxrwx@  1 silver  staff  2907296 Oct  2 18:47 demonsaw_cli
> drwxrwxrwx@  8 silver  staff      272 Oct  2 17:44 theme

**Comment by camgaertner** [comment]
Just wanted to add FTR that I have gotten Demonsaw 3.1.0 working on a MacBook with Sierra 10.12 by doing the steps I originally commented. 

You **should** be able to have full functionality of Demonsaw after confirming write access to the folder that **contains** `demonsaw.app`, like Eijah said. The directory is probably
`/Applications/demonsaw_osx_64`

If that doesn't work, the steps I used are:

```
cd /Applications/demonsaw_osx_64/demonsaw.app/Contents/MacOS/
sudo ./demonsaw
(enter your password)
```

However, running programs from third parties as root is not necessarily preferable unless debugging a trusted application.

**Comment by silvether** [comment]
So, I am now able to write a .TOML file and am back to my original bug outline. After deleting the app and starting fresh, I hit an issue that I had bypassed due to how I installed Demonsaw. The trouble was with the inability to write a .TOML due to a new feature in Gatekeeper(headache) called `'Gatekeeper Path Randomization'` . The feature launches an application in a read-only sandbox if the app is launched from its original folder/location and is not moved, this also does not play well with apps that are not signed with a valid Apple dev profile. Apple has also removed the ability to set your security settings to 'Anywhere' instead of 'App Store' or 'App Store and identified developers.' 

Here are logs showing a read-only sandbox if the app is launched from within the original folder in the archive:

`SecTranslocateCreateSecureDirectoryForURL: created /private/var/folders/77/1t_hk2xd1kn3vplds17ytmbm0000gn/T/AppTranslocation/A97EAA48-FCEB-404E-A271-4EA9AE10D4C1/d/demonsaw.app`

It also has a problem deleting the app from the sandbox:

`SecTranslocate: problem deleting translocation after app death: /private/var/folders/77/1t_hk2xd1kn3vplds17ytmbm0000gn/T/AppTranslocation/A97EAA48-FCEB-404E-A271-4EA9AE10D4C1/d/demonsaw.app`

Even with my regained ability to write the.TOML file, just like before, it still reverts the .TOML when the app is closed. 

Another thing to note, is that if you launch Demonsaw with the binary found in the .app bundle it will write the .TOML inside of the .app with root ownership and permissions. This .TOML is also reset after close.

`sudo ./demonsaw` ran from `/Applications/demonsaw.app/Contents/MacOS/`

Even running the more appropriate command of 'open' yields similar results, but instead writing to the root dir where the .app lives:

`sudo open demonsaw.app`

You will note that the .TOML is written as root and also is reset to default upon close.

The issue here does not appear to be root level privileges but what is happening to the .TOML when Demonsaw is closed. While Demonsaw is running the .TOML file shows the settings correctly.

**I am still stuck with the behaviour detailed in Steps 4-6 in of original post.**

##
**Can you confirm that the settings written to the .TOML during runtime are indeed perserved after closing the app, under both root and non-root users?**
##

_On a side note, there is a way to restore the "Anywhere" option for Gatekeeper via the terminal:_ 

_`sudo spctl --master-disable`_

_I will be testing combinations of these setting to see how they affect the outcome._

**Comment by silvether** [comment]
So after further testing, I was able to get Demonsaw to keep the configuration and not overwrite it when closed. Apparently there are two separate behaviours depending on how the app is closed?
1. Closing the app with the red X

<img width="76" alt="screen shot 2016-10-06 at 10 07 22 pm" src="https://cloud.githubusercontent.com/assets/19334356/19178596/55d56d32-8c11-11e6-8814-f5c94787ed0d.png">

**This method preserves the TOML as expected.**
1. Closing the app with the Quit option in the menu(Command - Q shortcut)

<img width="211" alt="screen shot 2016-10-06 at 10 07 33 pm" src="https://cloud.githubusercontent.com/assets/19334356/19178612/86f46f9e-8c11-11e6-92e4-dfbb7af3febc.png">

**This method resets the TOML to default on close. Sad panda..**.

Method 2 is the most common method(I have noted) of quitting an application in OSX(especially when fullscreen), and is the reason I overlooked it as a potential variable. We should work to ensure both methods yield the same outcome.

Changing the title of this bug to better reflect the current situation.
